pkg/k8s/controllers/errors.go

```diff
@@ -0,0 +1,35 @@
+package controllers
+
+// wrappedError is a helper to define PropagateError and PermanentError
+type wrappedError struct {
+	Original error
+}
+
+// Error delegates error interface impl to the Original
+func (we wrappedError) Error() string {
+	return we.Original.Error()
+}
+
+// PermanentError wraps an existing error to make it permanent (e.g. no more retries allowed), use
+// Permanent method to create it.
+type PermanentError struct {
+	wrappedError
+}
+
+// Permanent wraps an existing error to make it permanent (e.g. no more retries allowed).
+func Permanent(err error) error {
+	return PermanentError{wrappedError: wrappedError{Original: err}}
+}
+
+// PropagateError wraps an existing error to mark it as propagatable to controller. If Reconciler's handler
+// returns wraps the error with this struct, Reconciler unwraps it and return it to the controller infra.
+// Use Propagate method to create it.
+type PropagateError struct {
```

Comment:
The recommended wrapping pattern is to to implement Unwrap (see [here](https://github.com/getoutreach/gobox/blob/ffba60cecdc843119e94ff39211558005d852333/pkg/orerr/orerr.go#L126)).

This will ensure this plays well with Go's [error treatment](https://go.dev/blog/go1.13-errors).

Also, I do not recommend exporting the concrete error types and instead exporting `IsXYZ` method (much like with [IsRetryable](https://github.com/getoutreach/gobox/blob/ffba60cecdc843119e94ff39211558005d852333/pkg/orerr/orerr.go#L136)).  This will allow changing the underlying implementation later.

In fact, could you just use IsRetryable to mark errors are retryable?  Do you even need the Propagate and Permanent error types?

Reply:
Oh, I am going to close this PR - code moved to services repo. I will apply comments in the new repo later, thanks!!